File: Python3.x/36-Valid_Sudoku.py
# 常规思路堆逻辑
# Runtime: 48 ms, faster than 88.39% of Python3 online submissions for Valid Sudoku.
# Memory Usage: 13.3 MB, less than 36.37% of Python3 online submissions for Valid Sudoku.
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        if self.check_row(board) and self.check_col(board) and self.check_array(board):
            return True
        else:
            return False

    def check_row(self, board):
        for i in range(9):
            row = ""
            for j in range(9):
                if board[i][j] not in row:
                    row += board[i][j]
                elif board[i][j] != '.':
                    return False
        return True

    def check_col(self, board):
        for i in range(9):
            col = ""
            for j in range(9):
                if board[j][i] not in col:
                    col += board[j][i]
                elif board[j][i] != '.':
                    return False
        return True

    def check_array(self, board):
        for row in range(0, 9, 3):
            for col in range(0, 9, 3):
                arr = ""
                for i in range(row, row + 3):
                    for j in range(col, col + 3):
                        if board[i][j] not in arr:
                            arr += board[i][j]
                        elif board[i][j] != '.':
                            return False
        return True


# set
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        tmp = set()
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    if (i, board[i][j]) in tmp or (board[i][j], j) in tmp or (i // 3, j // 3, board[i][j]) in tmp:
                        return False
                    tmp.add((i, board[i][j]))
                    tmp.add((board[i][j], j))
                    tmp.add((i // 3, j // 3, board[i][j]))
        return True

File: Algorithms/C++/36-Valid_Sudoku.cpp
// Runtime: 12 ms, faster than 89.79% of C++ online submissions for Valid Sudoku.
// Memory Usage: 10 MB, less than 56.41% of C++ online submissions for Valid Sudoku.
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // check every column
        for (int i = 0; i < board.size(); ++i) {
            vector<char> visited;
            for (int j = 0; j < board[0].size(); ++j) {
                if (board[i][j] != '.')
                    if (find(visited.begin(), visited.end(), board[i][j]) != visited.end())
                        return false;
                    else
                        visited.push_back(board[i][j]);
            }
        }
        // check every row
        for (int j = 0; j < board[0].size(); ++j) {
            vector<char> visited;
            for (int i = 0; i < board[0].size(); ++i) {
                if (board[i][j] != '.')
                    if (find(visited.begin(), visited.end(), board[i][j]) != visited.end())
                        return false;
                    else
                        visited.push_back(board[i][j]);
            }
        }
        // check 3 x 3 sub-boxes
        for (int row = 0; row < 3; ++row) { // sub-box row
            for (int col = 0; col < 3; ++col) { // sub-box col
                vector<char> visited;
                for (int i = row * 3; i < row * 3 + 3; ++i)
                    for (int j = col * 3; j < col * 3 + 3; ++j)
                        if (board[i][j] != '.')
                            if (find(visited.begin(), visited.end(), board[i][j]) != visited.end())
                                return false;
                            else
                                visited.push_back(board[i][j]);
            }
        }
        return true;
    }
};
